@here/harp-mapview/lib/BoundsGenerator.ts

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright (C) 2017-2020 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import {
+    EarthConstants,
+    GeoCoordinates,
+    mercatorTilingScheme,
+    ProjectionType,
+    TileKey
+} from "@here/harp-geoutils";
+import { Frustum, Line3, Matrix4, Plane, Vector2, Vector3 } from "three";
+
+import { TileGeometryCreator } from "./geometry/TileGeometryCreator";
+import { MapView } from "./MapView";
+import { MapViewUtils } from "./Utils";
+
+export class BoundsGenerator {
+    //TODO: check what we really need here
+    constructor(private m_mapView: MapView) {}
+
+    generate(): GeoCoordinates[] {
+        //TODO: support spherical projection
+
+        //!!!!!!!ALTITUDE IS NOT TAKEN INTO ACCOUNT!!!!!!!!!
+        const geoPolygon: GeoCoordinates[] = [];
+
+        //CASE A: FLAT PROJECTION
+        if (this.m_mapView.projection.type === ProjectionType.Planar) {
+            // 1.) Raycast into all for corners of the canvas
+            //     => if an intersection is found, add it to the polygon
+            this.raycastCanvasCorners(geoPolygon);
+
+            // => All 4 corners found an intersection, therefore the screen is covered with the map
+            // and the polygon complete
+            if (geoPolygon.length === 4) {
+                return geoPolygon;
+            }
+
+            //2.) Raycast into the two corners of the horizon cutting the canvas sides
+            //    => if intersection is found, add it to the polygon
+            this.rayCastOnHorizonCorners(geoPolygon);
+
+            //Setup the frustum for further checks
+            const frustum = new Frustum().setFromProjectionMatrix(
+                new Matrix4().multiplyMatrices(
+                    this.m_mapView.camera.projectionMatrix,
+                    this.m_mapView.camera.matrixWorldInverse
+                )
+            );
+
+            // Setup the world corners for further checks
+            const worldCorners = TileGeometryCreator.instance.generateTilePlaneCorners(
+                mercatorTilingScheme.getGeoBox(new TileKey(0, 0, 0)),
```

Comment:
Is mercatorTilingScheme correct if projection != webMercator?
I guess you should use projection.worldExtent instead

Reply:
good catch, yes, probably there couldbe differences in the bounding box, it should use the mapView.projection